src/hyperlink/_url.py

```diff
@@ -882,15 +882,28 @@ def __init__(
         self._rooted = _typecheck("rooted", rooted, bool)
         self._userinfo = _textcheck("userinfo", userinfo, '/?#@')
 
-        uses_netloc = scheme_uses_netloc(self._scheme, uses_netloc)
+        if uses_netloc is None:
+            uses_netloc = scheme_uses_netloc(self._scheme, uses_netloc)
         self._uses_netloc = _typecheck("uses_netloc",
                                        uses_netloc, bool, NoneType)
-        # fixup for rooted consistency
-        if self._host:
+        will_have_authority = (
+            self._host or
+            (self._port and self._port != SCHEME_PORT_MAP.get(scheme))
+        )
+        if will_have_authority:
+            # fixup for rooted consistency; if there's any 'authority'
+            # represented in the textual URL, then the path must be rooted, and
+            # we're definitely using a netloc (there must be a ://).
             self._rooted = True
-        if (not self._rooted) and self._path and self._path[0] == '':
+            self._uses_netloc = True
+        if (not self._rooted) and self.path[:1] == (u'',):
             self._rooted = True
             self._path = self._path[1:]
+        if not will_have_authority and self._path and not self._rooted:
+            # If, after fixing up the path, there *is* a path and it *isn't*
+            # rooted, then we are definitely not using a netloc; if we did, it
+            # would make the path (erroneously) look like a hostname.
+            self._uses_netloc = False
```

Comment:
[This conversation](https://github.com/python-hyper/hyperlink/issues/24#issuecomment-312569915) makes me think that if `uses_netloc ` is not `None`, you should be comparing `self._uses_netloc` against `uses_netloc` and raising an exception if they do not match.  Same for `self._rooted`.

Reply:
I don't entirely disagree, but if we do that I'd like to keep that in a separate change, because that pushes us quite far down the direction of "`uses_netloc` is a bad idea and should not have been added", which I am increasingly believing is actually what we should be doing with this value.  Rather than raising in specific circumstances which you can't know about in advance because you don't and can't know what other code has done global cache-mangling with `register_scheme`.  So I there's a lot to discuss here beyond enforcing correctness, which is what this branch is trying to do.

Mostly folks just shouldn't be using this feature though.